File: cloud_functions/functions/src/export_user_data/models/export_user_data_request.ts
import { firestore } from 'firebase-admin';
import { UserId } from '../../user/models/user_id';
import { MapData } from '../../utils/map_converter';

export class ExportUserDataRequest {
    public id: string;
    public userId: UserId;
    public requestTime: firestore.Timestamp;
    public status: 'loading' | 'successful' | 'error';
    public downloadUrl: string | null;
    public totalBytes: number | null;
    public error: boolean | null;

    constructor(params: {
        id: string,
        userId: UserId,
        requestTime: firestore.Timestamp,
        downloadUrl: string | null,
        totalBytes: number | null,
        error: boolean | null,
        status: 'loading' | 'successful' | 'error',
    }) {
        this.id = params.id;
        this.userId = params.userId;
        this.requestTime = params.requestTime;
        this.downloadUrl = params.downloadUrl;
        this.totalBytes = params.totalBytes;
        this.error = params.error;
        this.status = params.status;
    }

    static fromData(id: string, data: MapData<any>): ExportUserDataRequest {
        return new ExportUserDataRequest({
            id: id,
            userId: new UserId(data['userId']),
            requestTime: data['requestTime'],
            downloadUrl: data['downloadUrl'],
            totalBytes: data['totalBytes'],
            error: data['error'],
            status: data['status'],
        });
    }
}

File: packages/export_user_data/export_user_data_models/lib/src/models/export_user_data_request.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:export_user_data_models/export_user_data_models.dart';
import 'package:export_user_data_models/src/models/export_user_data_request_status.dart';
import 'package:filesize_ns/filesize_ns.dart';

class ExportUserDataRequest {
  final String id;
  final String userId;
  final DateTime requestTime;
  final DateTime? expiresOn;
  final String? downloadUrl;
  final int? totalBytes;
  final String? error;
  final ExportUserDataRequestStatus status;

  const ExportUserDataRequest._({
    required this.id,
    required this.userId,
    required this.requestTime,
    required this.expiresOn,
    required this.downloadUrl,
    required this.totalBytes,
    required this.status,
    required this.error,
  });

  factory ExportUserDataRequest.createForUserId(
      String requestId, String userId) {
    return ExportUserDataRequest._(
      id: requestId,
      userId: userId,
      requestTime: DateTime.now(),
      expiresOn: null,
      downloadUrl: null,
      totalBytes: null,
      status: ExportUserDataRequestStatus.loading,
      error: null,
    );
  }

  factory ExportUserDataRequest.fromData(
      String key, Map<String, dynamic> data) {
    try {
      return ExportUserDataRequest._(
        id: key,
        userId: data['userId'],
        requestTime: (data['requestTime'] as Timestamp).toDate(),
        expiresOn: data['expiresOn'] != null
            ? (data['expiresOn'] as Timestamp).toDate()
            : null,
        downloadUrl: data['downloadUrl'],
        totalBytes: data['totalBytes'],
        status: ExportUserDataRequestStatusConverter.fromData(data['status']),
        error: data['error'],
      );
    } catch (e) {
      throw Exception('ExportUserDataRequest Parsing Exception!');
    }
  }

  Map<String, dynamic> toData() {
    return {
      'userId': userId,
      'requestTime': Timestamp.fromDate(requestTime),
      'expiresOn': expiresOn != null ? Timestamp.fromDate(expiresOn!) : null,
      'downloadUrl': downloadUrl,
      'totalBytes': totalBytes,
      'status': ExportUserDataRequestStatusConverter.toData(status),
      'error': error,
    };
  }

  ExportUserDataRequestView toView() {
    return ExportUserDataRequestView(
      userId: userId,
      requestTime: requestTime,
      expiresOn: expiresOn,
      downloadUrl: downloadUrl,
      bytesSize: totalBytes != null ? filesize(totalBytes) : null,
      status: status,
      error: error,
    );
  }
}
